src/rust/ide/src/controller/visualisation.rs

```diff
@@ -0,0 +1,166 @@
+//! Visualisation controller.
+//!
+//! Ths Visualisation Controller is Responsible identifying all the available visualisation natively
+//! embedded in IDE and available within the project's `visualisation` folder. The
+//! `Visualisation Controller` lives as long as the `Project Controller`.
+
+use crate::prelude::*;
+
+use std::rc::Rc;
+use enso_protocol::language_server;
+
+
+
+// =============
+// === Error ===
+// =============
+
+/// Enumeration of errors used in `Visualisation Controller`.
+#[derive(Debug,Fail)]
+#[allow(missing_docs)]
+pub enum VisualiserError {
+    #[fail(display = "Visualiser \"{}\" not found.", identifier)]
+    NotFound {
+        identifier : VisualiserIdentifier
+    }
+}
+
+
+
+// ===============================
+// === VisualisationIdentifier ===
+// ===============================
+
+/// This enum is used to identify a visualiser both in the project folder or natively embedded in
+/// IDE.
+#[derive(Debug,Clone,PartialEq,Eq,Display)]
+#[allow(missing_docs)]
+pub enum VisualiserIdentifier {
+    Embedded(String),
+    File(language_server::Path)
+}
+
+
+
+// ==============
+// === Handle ===
+// ==============
+
+/// Visualisation Controller's state.
+#[derive(Debug,Clone)]
+pub struct Handle {
+    language_server_rpc  : Rc<language_server::Connection>,
+    embedded_visualisers : HashMap<String,String>
+}
+
+impl Handle {
+    /// Creates a new visualisation controller.
+    pub fn new
+    ( language_server_rpc  : Rc<language_server::Connection>
+    , embedded_visualisers : HashMap<String,String>) -> Self {
+        Self {language_server_rpc,embedded_visualisers}
+    }
+
+    /// Get a list of all available visualisers.
+    pub async fn list_visualisers(&self) -> FallibleResult<Vec<VisualiserIdentifier>> {
+        let root_id    = self.language_server_rpc.content_root();
+        let vis_path   = language_server::Path{root_id,segments:vec!["visualisation".into()]};
+        let file_list  = self.language_server_rpc.file_list(vis_path).await?;
+        let result     = self.embedded_visualisers.keys();
+        let result     = result.map(|identifier| VisualiserIdentifier::Embedded(identifier.clone()));
+        let mut result = result.collect::<Vec<VisualiserIdentifier>>();
+        result.extend(file_list.paths.iter().filter_map(|object| {
+            if let language_server::FileSystemObject::File{name,path} = object {
+                let mut path = path.clone();
+                path.segments.push(name.to_string());
+                let root_id                  = path.root_id;
+                let segments                 = path.segments;
+                let path                     = language_server::Path{root_id,segments};
+                let visualisation_identifier = VisualiserIdentifier::File(path);
+                Some(visualisation_identifier)
+            } else {
+                None
+            }
+        }));
```

Comment:
A minor thing, regarding iterators: instead of creating vector and then extending it, you can use `chain` method to chain two iterators and do then the final collecting. That in theory is better, because we do only one allocation of vector.